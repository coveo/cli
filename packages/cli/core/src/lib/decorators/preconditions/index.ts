// TODO: Remove, unusued
export * from './git';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './node';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npx';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npm';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './ng';
