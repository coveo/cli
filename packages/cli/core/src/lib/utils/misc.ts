import {resolve} from 'path';
import {valid} from 'semver';

export function getPackageVersion(packageName: string) {
  const pathToPackageJson = resolve(
    __dirname,
    ...new Array(3).fill('..'),
    'package.json'
  );
  const pkg = require(pathToPackageJson);
  const dep =
    pkg.dependencies[packageName] ||
    pkg.devDependencies[packageName] ||
    pkg.peerDependencies[packageName];
  const defaultVersion = valid(dep?.toString())!;

  return defaultVersion;
}
