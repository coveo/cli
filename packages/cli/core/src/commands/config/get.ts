import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {Config} from '@coveo/cli-commons/config/config';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Get extends CLICommand {
  public static description = 'Display the current Coveo CLI configuration.';

  public static args = [
    {
      name: 'key',
      description: 'The config key for which to show the value',
      required: false,
    },
  ];

  public static examples: Example[] = [
    {
      command: 'coveo config:get',
      description: 'Get all the configuration values',
    },
    {
      command: 'coveo config:get organization',
      description: 'Get the organization to which you are connected',
    },
    {
      command: 'coveo config:get accessToken',
      description: 'Get the access token given to you by the Coveo Platform',
    },
  ];

  @Trackable()
  public async run() {
    const {args} = await this.parse(Get);
    const cfg = new Config(this.config.configDir);
    const keysToRender = args.key ? [args.key] : undefined;
    ConfigRenderer.render(cfg, keysToRender);
  }
}
