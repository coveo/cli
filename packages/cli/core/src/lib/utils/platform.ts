import {Config} from '@coveo/cli-commons/config/config';

export function getTargetOrg(config: Config, target?: string) {
  if (target) {
    return target;
  }
  const cfg = config.get();
  return cfg.organization;
}
