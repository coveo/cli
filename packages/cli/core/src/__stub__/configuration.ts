import {Region} from '@coveo/platform-client';
import type {Config, Configuration} from '@coveo/cli-commons/config/config';
import {PlatformEnvironment} from '@coveo/cli-commons/platform/environment';

export const defaultConfiguration = {
  environment: PlatformEnvironment.Dev,
  organization: 'my-org',
  region: Region.US,
  accessToken: 'some-api-key-or-token',
  anonymous: false,
} as Configuration;

export const configurationMock: (
  configuration?: Configuration
) => () => Config =
  (configuration = defaultConfiguration) =>
  () =>
    ({
      get: () => configuration,
    } as Config);
