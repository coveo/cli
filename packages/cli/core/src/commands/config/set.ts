import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {CliUx, Flags} from '@oclif/core';
import {Config} from '@coveo/cli-commons/config/config';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {
  IsAuthenticated,
  Preconditions,
} from '@coveo/cli-commons/preconditions/index';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import {InvalidCommandError} from '../../lib/errors/InvalidCommandError';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Set extends CLICommand {
  public static description = 'Modify the current Coveo CLI configuration.';

  public static flags = {
    // TODO CDX-1246
    environment: Flags.string({
      char: 'e',
      hidden: true,
    }),
    // TODO CDX-1246
    region: Flags.string({
      char: 'r',
      hidden: true,
    }),
    organization: Flags.string({
      char: 'o',
      description:
        'The identifier of the organization inside which to perform operations. See <https://docs.coveo.com/en/1562/#organization-id-and-other-information>.',
      helpValue: 'myOrgID',
    }),
  };

  public static examples: Example[] = [
    {
      command: 'coveo config:set --organization myOrgId',
      description: 'connect to a different organization',
    },
  ];

  @Trackable()
  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = await this.parse(Set);
    // TODO CDX-1246
    if (flags.environment || flags.region) {
      CliUx.ux.error(
        'To connect to a different region or environment, use the `auth:login` command'
      );
    }
    if (Object.entries(flags).length === 0) {
      throw new InvalidCommandError('Command should contain at least 1 flag');
    }
    const cfg = new Config(this.config.configDir);
    if (flags.organization) {
      await this.verifyOrganization(flags.organization);
      cfg.set('organization', flags.organization);
    }

    ConfigRenderer.render(cfg);
  }

  private async verifyOrganization(org: string) {
    const hasAccess = await new AuthenticatedClient().getUserHasAccessToOrg(
      org
    );
    if (!hasAccess) {
      this.error(
        `You either don't have access to organization ${org}, or it doesn't exist.`
      );
    }
  }
}
