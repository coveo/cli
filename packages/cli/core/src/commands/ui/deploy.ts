import {validate, Schema} from 'jsonschema';
import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {
  AuthenticationType,
  HasNecessaryCoveoPrivileges,
  IsAuthenticated,
  Preconditions,
} from '@coveo/cli-commons/preconditions/index';
import type {HostedPage, New} from '@coveo/platform-client';
import type PlatformClient from '@coveo/platform-client';
import {createSearchPagesPrivilege} from '@coveo/cli-commons/preconditions/platformPrivilege';
import {Flags} from '@oclif/core';
import {readJsonSync, ensureDirSync, readFileSync} from 'fs-extra';
import {DeployConfigError} from '../../lib/errors/deployErrors';
import {join} from 'path';
import {Example} from '@oclif/core/lib/interfaces';
import {confirm, startSpinner, stopSpinner} from '@coveo/cli-commons/utils/ux';
import {getTargetOrg} from '../../lib/utils/platform';
import {Config} from '@coveo/cli-commons/config/config';
import {organization} from '../../lib/flags/platformCommonFlags';
import dedent from 'ts-dedent';

interface FileInput {
  path: string;
}

interface JavaScriptFileInput extends FileInput {
  isModule: boolean;
}

export interface DeployConfig {
  name: string;
  dir: string;
  htmlEntryFile: FileInput;
  javascriptEntryFiles?: JavaScriptFileInput[];
  javascriptUrls?: JavaScriptFileInput[];
  cssEntryFiles?: FileInput[];
  cssUrls?: FileInput[];
  schemaVersion: string;
}

const FileInputSchema: Schema = {
  type: 'object',
  required: ['path'],
  properties: {
    path: {type: 'string'},
  },
};

const JavaScriptFileInputSchema: Schema = {
  type: 'object',
  required: ['path', 'isModule'],
  properties: {
    path: {type: 'string'},
    patisModuleh: {type: 'boolean'},
  },
};

const DeployConfigSchema: Schema = {
  type: 'object',
  required: ['name', 'dir', 'htmlEntryFile', 'schemaVersion'],
  properties: {
    name: {type: 'string'},
    dir: {type: 'string'},
    htmlEntryFile: FileInputSchema,
    javascriptEntryFiles: {type: 'array', items: JavaScriptFileInputSchema},
    javascriptUrls: {type: 'array', items: JavaScriptFileInputSchema},
    cssEntryFiles: {type: 'array', items: FileInputSchema},
    cssUrls: {type: 'array', items: FileInputSchema},
    schemaVersion: {type: 'string'},
  },
};

export default class Deploy extends CLICommand {
  public static summary =
    'Deploy your search application to the Coveo infrastructure.';

  public static description = `The target "config" must contain the following parameters:
  {
    "name": "The name of the hosted search page.",
    "dir": "The directory of the hosted search page.",
    "htmlEntryFile": {
      "path": "The path to an HTML file containing the HTML markup of the hosted page."
    },
    "javascriptEntryFiles": [
      {
        "path": "The path to a bundled Javascript file.",
        "isModule": "Whether the inline code should be treated as a JavaScript module. If this property is true, the type property will be set to "module" on the script tag."
      }
    ],
    "javascriptUrls": [
      {
        "path": "The URL of the JavaScript source file.",
        "isModule": "Whether the inline code should be treated as a JavaScript module. If this property is true, the type property will be set to "module" on the script tag."
      }
    ],
    "cssEntryFiles": [
      {
        "path": "The path to a bundled CSS file."
      }
    ],
    "cssUrls": [
      {
        "path": "The URL of the CSS stylesheet."
      }
    ]
  }`;

  public static aliases: string[] = ['atomic:deploy'];

  public static flags = {
    pageId: Flags.string({
      char: 'p',
      description: 'The existing ID of the target Hosted search page.',
      helpValue: '7944ff4a-9943-4999-a3f6-3e81a7f6fb0a',
      required: false,
    }),
    config: Flags.string({
      char: 'c',
      description: 'The path to the deployment JSON configuration.',
      helpValue: 'coveo.deploy.json',
      default: 'coveo.deploy.json',
      required: false,
    }),
    ...organization(
      'The unique identifier of the organization where to deploy the hosted page.'
    ),
  };

  public static examples: Example[] = [
    {
      command: '<%= config.bin %> <%= command.id %>',
      description:
        'Create a new Hosted Page according to the configuration in the file "coveo.deploy.json"',
    },
    {
      command:
        '<%= config.bin %> <%= command.id %> -p 7944ff4a-9943-4999-a3f6-3e81a7f6fb0a',
      description:
        'Update the Hosted Page whose ID is "7944ff4a-9943-4999-a3f6-3e81a7f6fb0a" according to the configuration in the file "coveo.deploy.json"',
    },
    {
      command: '<%= config.bin %> <%= command.id %> -c ./configs/myconfig.json',
      description:
        'Create a new Hosted Page according to the configuration in the file located at "./configs/myconfig.json"',
    },
  ];

  private flags!: {organization?: string; pageId?: string; config: string};
  private organization!: string;
  private platformClient!: PlatformClient;
  private hostedPage!: New<HostedPage>;
  private pageId: string | undefined;

  @Trackable()
  @Preconditions(
    IsAuthenticated([AuthenticationType.OAuth, AuthenticationType.ApiKey]),
    HasNecessaryCoveoPrivileges(createSearchPagesPrivilege)
  )
  public async run() {
    await this.computeFlags();
    this.computeOrgId();
    await this.instantiatePlatformClient();
    this.computeHostedPage();
    await this.computePageId();

    if (this.pageId) {
      await this.updatePage({...this.hostedPage, id: this.pageId});
    } else {
      await this.createPage(this.hostedPage);
    }
  }

  private async computeFlags() {
    this.flags = (await this.parse(Deploy)).flags;
  }

  private computeOrgId() {
    this.organization = getTargetOrg(
      this.configuration,
      this.flags.organization
    );
  }

  private async instantiatePlatformClient() {
    this.platformClient = await new AuthenticatedClient().getClient({
      organization: this.organization,
    });
  }

  private computeHostedPage() {
    const deployConfigPath = this.flags.config;
    const deployConfig: DeployConfig =
      readJsonSync(deployConfigPath, {throws: false}) || {};
    this.validateDeployConfigJson(deployConfigPath, deployConfig);
    this.hostedPage = this.buildHostedPage(deployConfig);
  }

  private async computePageId() {
    this.pageId = this.getPageIdFromFlags();

    if (this.pageId) {
      return;
    }

    this.pageId = await this.getPageIdFromHostedPage();

    if (this.pageId) {
      await this.confirmOverwrite();
    }
  }

  private getPageIdFromFlags(): string | undefined {
    return this.flags.pageId;
  }

  private async getPageIdFromHostedPage(): Promise<string | undefined> {
    const response = await this.platformClient.hostedPages.list({
      filter: this.hostedPage.name,
    });
    return response.items.find((page) => page.name === this.hostedPage.name)
      ?.id;
  }

  private async confirmOverwrite(): Promise<void | never> {
    if (
      !(await confirm(
        dedent`
      Overwrite page with ID: "${this.pageId}" in organization ${this.organization}? (y/n)`,
        false
      ))
    ) {
      this.error(
        'Page name must be unique, try changing the "name" field in "coveo.deploy.json".'
      );
    }
  }

  private validateDeployConfigJson(
    deployConfigPath: string,
    config: DeployConfig
  ) {
    const validation = validate(config, DeployConfigSchema);
    if (!validation.valid) {
      throw new DeployConfigError(deployConfigPath, validation.errors);
    }
  }

  private buildHostedPage({
    dir,
    htmlEntryFile,
    name,
    cssEntryFiles = [],
    cssUrls = [],
    javascriptEntryFiles = [],
    javascriptUrls = [],
  }: DeployConfig): New<HostedPage> {
    ensureDirSync(dir);

    return {
      name,
      html: readFileSync(join(dir, htmlEntryFile.path), 'utf-8'),
      javascript: [
        ...javascriptEntryFiles.map(({path, isModule}) => ({
          inlineContent: readFileSync(join(dir, path), 'utf-8'),
          isModule,
        })),
        ...javascriptUrls.map(({path, isModule}) => ({
          url: path,
          isModule,
        })),
      ],
      css: [
        ...cssEntryFiles.map(({path}) => ({
          inlineContent: readFileSync(join(dir, path), 'utf-8'),
        })),
        ...cssUrls.map(({path}) => ({
          url: path,
        })),
      ],
    };
  }

  private async createPage(page: New<HostedPage>) {
    startSpinner(`Creating new Hosted Page named "${page.name}".`);

    const response = await this.platformClient.hostedPages.create(page);
    stopSpinner();
    this.log(
      `To update your page, run "coveo ${this.identifier} -p=${response.id}".`
    );
  }

  private async updatePage(page: HostedPage) {
    startSpinner(`Updating existing Hosted Page with the id "${page.id}".`);

    await this.platformClient.hostedPages.update(page);
    stopSpinner();
  }

  private get configuration() {
    return new Config(this.config.configDir);
  }
}
