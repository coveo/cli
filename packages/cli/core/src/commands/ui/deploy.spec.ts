jest.mock('child_process');
jest.mock('fs-extra');
jest.mock('jsonschema');

jest.mock('@coveo/cli-commons/utils/ux');
jest.mock('@coveo/cli-commons/config/config');
jest.mock('@coveo/cli-commons/preconditions/trackable');
jest.mock('@coveo/cli-commons/preconditions/authenticated');
jest.mock('@coveo/cli-commons/preconditions/apiKeyPrivilege');
jest.mock('@coveo/cli-commons/platform/authenticatedClient');
jest.mock('@coveo/platform-client');

jest.mock('../../lib/utils/process');
jest.mock('../../lib/oauth/oauth');
jest.mock('../../lib/utils/misc');

import {test} from '@oclif/test';
import {spawnProcess} from '../../lib/utils/process';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import PlatformClient, {
  CoveoPlatformClientError,
  HostedPage,
  New,
} from '@coveo/platform-client';
import {Config, Configuration} from '@coveo/cli-commons/config/config';
import {
  HasNecessaryCoveoPrivileges,
  IsAuthenticated,
} from '@coveo/cli-commons/preconditions/index';
import {getPackageVersion} from '../../lib/utils/misc';
import {configurationMock} from '../../__stub__/configuration';
import {mockPreconditions} from '@coveo/cli-commons/preconditions/mockPreconditions';
import {readJSONSync, readFileSync} from 'fs-extra';
import {DeployConfig} from './deploy';
import {validate, ValidatorResult} from 'jsonschema';
import {join} from 'path';
import {confirm} from '@coveo/cli-commons/utils/ux';

const actualValidate = jest.requireActual('jsonschema').validate;

type DeepPartial<T> = T extends Function
  ? T
  : T extends object
  ? {[P in keyof T]?: DeepPartial<T[P]>}
  : T;

const validJsonConfig: DeployConfig = {
  name: 'my page',
  dir: 'dist',
  htmlEntryFile: {
    path: 'index.html',
  },
  javascriptEntryFiles: [
    {
      path: 'index.js',
      isModule: false,
    },
  ],
  javascriptUrls: [
    {
      path: 'https://static.cloud.coveo.com/atomic/v2/atomic.esm.js',
      isModule: true,
    },
  ],
  cssEntryFiles: [
    {
      path: 'css/index.css',
    },
  ],
  cssUrls: [
    {
      path: 'https://static.cloud.coveo.com/atomic/v2/themes/coveo.css',
    },
  ],
  schemaVersion: '1.0.0',
};

const pageTestId = 'thisistheid';
describe('ui:deploy', () => {
  const mockedConfig = jest.mocked(Config);
  const mockedSpawnProcess = jest.mocked(spawnProcess);
  const mockedPlatformClient = jest.mocked(PlatformClient);
  const mockedGetPackageVersion = jest.mocked(getPackageVersion);
  const mockedAuthenticatedClient = jest.mocked(AuthenticatedClient);
  const mockedApiKeyPrivilege = jest.mocked(HasNecessaryCoveoPrivileges);
  const mockedIsAuthenticated = jest.mocked(IsAuthenticated);
  const mockedReadJSON = jest.mocked(readJSONSync);
  const mockedReadFileSync = jest.mocked(readFileSync);
  const mockedValidate = jest.mocked(validate);
  const mockHostedPageCreate = jest.fn();
  const mockHostedPageList = jest.fn();
  const mockHostedPageUpdate = jest.fn();
  const mockedConfirm = jest.mocked(confirm);

  const preconditionStatus = {
    apiKey: true,
    authentication: true,
  };

  const doMockPreconditions = function () {
    preconditionStatus.apiKey = true;
    preconditionStatus.authentication = true;
    const mockedPreconditions = mockPreconditions(preconditionStatus);
    mockedApiKeyPrivilege.mockReturnValue(mockedPreconditions.apiKey);
    mockedIsAuthenticated.mockReturnValue(mockedPreconditions.authentication);
  };

  const doMockSpawnProcess = () => {
    mockedSpawnProcess.mockReturnValue(Promise.resolve(0));
  };

  const doMockedGetPackageVersion = () => {
    mockedGetPackageVersion.mockReturnValue('1.0.0');
  };

  const doMockedValidReadJSON = () => {
    mockedReadJSON.mockReturnValue(validJsonConfig);
  };

  const doMockedReadFileSync = () => {
    mockedReadFileSync.mockReturnValue('somefilecontents');
  };

  const doMockValidate = (valid = true) =>
    mockedValidate.mockReturnValue({valid} as ValidatorResult);

  const doActualValidate = () =>
    mockedValidate.mockImplementationOnce((...args) => actualValidate(...args));

  const doMockConfiguration = () => {
    mockedConfig.mockImplementation(
      configurationMock({
        accessToken: 'foo',
        environment: 'dev',
        region: 'us',
        organization: 'my-org',
      } as Configuration)
    );
  };

  const doMockAuthenticatedClient = () => {
    mockedAuthenticatedClient.mockImplementation(
      () =>
        ({
          createImpersonateApiKey: jest.fn(),
          getUsername: () => Promise.resolve('bob@coveo.com'),
          getClient: () =>
            Promise.resolve(
              mockedPlatformClient.getMockImplementation()!({
                accessToken: 'foo',
                organizationId: 'my-org',
              })
            ),
          cfg: mockedConfig.getMockImplementation()!('./'),
        } as unknown as AuthenticatedClient)
    );
  };
  const doMockPlatformClient = () => {
    mockedPlatformClient.mockImplementation(
      () =>
        ({
          initialize: () => Promise.resolve(),
          hostedPages: {
            create: mockHostedPageCreate.mockResolvedValue({id: 'somePageId'}),
            update: mockHostedPageUpdate.mockResolvedValue({}),
            list: mockHostedPageList.mockResolvedValue({
              items: [],
            }),
          },
        } as unknown as PlatformClient)
    );
  };

  let modifiedConfig: DeepPartial<DeployConfig>;

  beforeAll(() => {
    mockedConfirm.mockResolvedValue(true);
  });

  beforeEach(() => {
    doMockedGetPackageVersion();
    doMockSpawnProcess();
    doMockPlatformClient();
    doMockConfiguration();
    doMockAuthenticatedClient();
    doMockPreconditions();
    doMockValidate();
    doMockedValidReadJSON();
    doMockedReadFileSync();
    modifiedConfig = JSON.parse(JSON.stringify(validJsonConfig));
  });

  afterEach(() => {
    mockedApiKeyPrivilege.mockClear();
  });

  describe('when preconditions are not respected', () => {
    test
      .do(() => {
        preconditionStatus.apiKey = false;
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/apiKey Precondition Error/)
      .it(
        'should not execute the command if the API key preconditions are not respected'
      );

    test
      .do(() => {
        preconditionStatus.authentication = false;
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/authentication Precondition Error/)
      .it(
        'should not execute the command if the preconditions are not respected'
      );
  });

  describe('validating JSON config file', () => {
    test
      .do(() => {
        mockedReadJSON.mockImplementationOnce(() => null);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/Parsing: coveo.deploy.json./)
      .it(
        'should exit with an error when the default config file does not exist'
      );

    test
      .do(() => {
        mockedReadJSON.mockImplementationOnce(() => null);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy', '-c=myconfig.json'])
      .catch(/Parsing: myconfig.json./)
      .it(
        'should exit with an error when the defined config file does not exist'
      );

    test
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .it(
        'should read the default config file without throwing when it exists & is valid',
        () => {
          expect(mockedReadJSON).toHaveBeenCalledWith(
            'coveo.deploy.json',
            expect.any(Object)
          );
        }
      );

    test
      .do(() => {
        delete modifiedConfig.htmlEntryFile;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/instance requires property "htmlEntryFile"/)
      .it(
        'should exit with an error when the htmlEntryFile is not part of the JSON config'
      );

    test
      .do(() => {
        delete modifiedConfig.schemaVersion;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/instance requires property "schemaVersion"/)
      .it(
        'should exit with an error when the schemaVersion is not part of the JSON config'
      );

    test
      .do(() => {
        delete modifiedConfig.dir;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/instance requires property "dir"/)
      .it(
        'should exit with an error when the dir is not part of the JSON config'
      );

    test
      .do(() => {
        delete modifiedConfig.cssEntryFiles![0]!.path;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/requires property "path"/)
      .it(
        'should exit with an error when the cssEntryFiles path is not part of the JSON config'
      );

    test
      .do(() => {
        delete modifiedConfig.cssUrls![0]!.path;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/requires property "path"/)
      .it(
        'should exit with an error when the cssUrls path is not part of the JSON config'
      );

    test
      .do(() => {
        delete modifiedConfig.javascriptEntryFiles![0]!.isModule;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/requires property "isModule"/)
      .it(
        'should exit with an error when the javascriptEntryFiles isModule is not part of the JSON config'
      );

    test
      .do(() => {
        delete modifiedConfig.javascriptEntryFiles![0]!.path;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/requires property "path"/)
      .it(
        'should exit with an error when the javascriptEntryFiles path is not part of the JSON config'
      );

    test
      .do(() => {
        delete modifiedConfig.javascriptUrls![0]!.isModule;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/requires property "isModule"/)
      .it(
        'should exit with an error when the javascriptUrls isModule is not part of the JSON config'
      );

    test
      .do(() => {
        delete modifiedConfig.javascriptUrls![0]!.path;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
        doActualValidate();
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .catch(/requires property "path"/)
      .it(
        'should exit with an error when the javascriptUrls path is not part of the JSON config'
      );
  });

  describe('fetching file contents defined in the config', () => {
    test
      .do(() => {
        delete modifiedConfig.cssEntryFiles;
        delete modifiedConfig.javascriptEntryFiles;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .it('should fetch the htmlEntryFile content', () => {
        expect(mockedReadFileSync).toHaveBeenCalledWith(
          join(validJsonConfig.dir, validJsonConfig.htmlEntryFile.path),
          'utf-8'
        );
      });

    test
      .do(() => {
        delete modifiedConfig.javascriptEntryFiles;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .it('should fetch the cssEntryFiles content', () => {
        expect(mockedReadFileSync).toHaveBeenNthCalledWith(
          2,
          join(validJsonConfig.dir, validJsonConfig.cssEntryFiles![0].path),
          'utf-8'
        );
      });

    test
      .do(() => {
        delete modifiedConfig.cssEntryFiles;
        mockedReadJSON.mockImplementationOnce(() => modifiedConfig);
      })
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .it('should fetch the javascriptEntryFiles content', () => {
        expect(mockedReadFileSync).toHaveBeenNthCalledWith(
          2,
          join(
            validJsonConfig.dir,
            validJsonConfig.javascriptEntryFiles![0].path
          ),
          'utf-8'
        );
      });
  });

  describe('interacting with Platform client', () => {
    const expectedHostedPage: New<HostedPage> = {
      html: 'somefilecontents',
      name: validJsonConfig.name,
      javascript: [
        ...validJsonConfig.javascriptEntryFiles!.map(({isModule}) => ({
          inlineContent: 'somefilecontents',
          isModule,
        })),
        ...validJsonConfig.javascriptUrls!.map(({isModule, path}) => ({
          url: path,
          isModule,
        })),
      ],
      css: [
        ...validJsonConfig.cssEntryFiles!.map(() => ({
          inlineContent: 'somefilecontents',
        })),
        ...validJsonConfig.cssUrls!.map(({path}) => ({
          url: path,
        })),
      ],
    };

    test
      .stdout()
      .stderr()
      .command(['ui:deploy'])
      .it(
        'should call hostedPages.list and hostedPages.create when no page id argument is passed',
        ({stdout, stderr}) => {
          expect(mockHostedPageList).toHaveBeenCalledWith(
            expect.objectContaining({
              filter: expectedHostedPage.name,
            })
          );
          expect(stdout).toMatchSnapshot();
          expect(stderr).toMatchSnapshot();
          expect(mockHostedPageCreate).toHaveBeenCalledWith(expectedHostedPage);
        }
      );

    test
      .stdout()
      .stderr()
      .do(() => {
        mockedConfirm.mockResolvedValueOnce(true);
        mockHostedPageList.mockResolvedValueOnce({
          items: [{name: validJsonConfig.name, id: pageTestId}],
        });
      })
      .command(['ui:deploy'])
      .it(
        'when no page id argument is passed and a page with the same name already exists, it should ask the user for confirmation',
        () => {
          expect(mockedConfirm).toBeCalled();
        }
      );

    test
      .stdout()
      .stderr()
      .do(() => {
        mockedConfirm.mockResolvedValueOnce(true);
        mockHostedPageList.mockResolvedValueOnce({
          items: [{name: validJsonConfig.name, id: pageTestId}],
        });
      })
      .command(['ui:deploy'])
      .it(
        'when no page id argument is passed, a page with the same name already exists, and the user confirms the overwrite, it should call hostedPages.update',
        () => {
          expect(mockHostedPageUpdate).toHaveBeenCalledWith({
            ...expectedHostedPage,
            id: pageTestId,
          });
        }
      );

    test
      .stdout()
      .stderr()
      .do(() => {
        mockedConfirm.mockResolvedValueOnce(false);
        mockHostedPageList.mockResolvedValueOnce({
          items: [{name: validJsonConfig.name, id: pageTestId}],
        });
      })
      .command(['ui:deploy'])
      .catch((err) => expect(err).toMatchSnapshot())
      .it(
        'when no page id argument is passed, a page with the same name already exists, the user declines the overwrite, it should not call hostedPages.update',
        () => {
          expect(mockHostedPageUpdate).not.toBeCalled();
        }
      );

    test
      .stdout()
      .stderr()
      .command(['ui:deploy', `-p=${pageTestId}`])
      .it(
        'when a page id argument is passed, it should call hostedPages.update',
        ({stdout, stderr}) => {
          expect(stdout).toMatchSnapshot();
          expect(stderr).toMatchSnapshot();
          expect(mockHostedPageUpdate).toHaveBeenCalledWith({
            ...expectedHostedPage,
            id: pageTestId,
          });
        }
      );

    test
      .stdout()
      .stderr()
      .do(() => {
        mockHostedPageUpdate.mockImplementationOnce(() => {
          const err = new CoveoPlatformClientError();
          err.title = 'SO_BAD';
          err.detail = 'this is bad';
          err.xRequestId = 'b33pb00p';
          err.status = 418;
          throw err;
        });
      })
      .command(['ui:deploy', `-p=${pageTestId}`])
      .catch((err) => expect(err).toMatchSnapshot())
      .it('should print an API Error if the update fails');
  });
});
