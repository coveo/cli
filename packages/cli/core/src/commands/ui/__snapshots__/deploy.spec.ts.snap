// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui:deploy interacting with Platform client should call hostedPages.list and hostedPages.create when no page id argument is passed 1`] = `
"To update your page, run "coveo ui:deploy -p=somePageId".
"
`;

exports[`ui:deploy interacting with Platform client should call hostedPages.list and hostedPages.create when no page id argument is passed 2`] = `""`;

exports[`ui:deploy interacting with Platform client should print an API Error if the update fails 1`] = `
[APIError: 
Status code: 418
Error code: [31mSO_BAD[39m
Message: [31mthis is bad[39m
Request ID: [31mb33pb00p[39m]
`;

exports[`ui:deploy interacting with Platform client when a page id argument is passed, it should call hostedPages.update 1`] = `""`;

exports[`ui:deploy interacting with Platform client when a page id argument is passed, it should call hostedPages.update 2`] = `""`;

exports[`ui:deploy interacting with Platform client when no page id argument is passed, a page with the same name already exists, the user declines the overwrite, it should not call hostedPages.update 1`] = `[CLI Error: Page name must be unique, try changing the "name" field in "coveo.deploy.json".]`;
